pkg/operator/csidriveroperator/driverstarter_hypershift.go

```diff
@@ -0,0 +1,231 @@
+	// the startup a lot
+	// - All necessary informers are populated.
+	// - CSO then just Run()s
+	// All ControllerManagers are not running at this point! They will be
+	// started in sync() when their platform is detected.
+	c.controllers = []csiDriverControllerManager{}
+	for _, cfg := range driverConfigs {
+		mgr, ctrlRelatedObjects := c.createCSIControllerManager(cfg, resyncInterval)
+		c.controllers = append(c.controllers, csiDriverControllerManager{
+			operatorConfig:     cfg,
+			mgr:                mgr,
+			running:            false,
+			ctrlRelatedObjects: ctrlRelatedObjects,
+		})
+	}
+
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(c.guestClient.OperatorClient).WithInformers(
+		guestClients.OperatorClient.Informer(),
+		guestClients.ConfigInformers.Config().V1().Infrastructures().Informer(),
+		guestClients.ConfigInformers.Config().V1().FeatureGates().Informer(),
+		guestClients.KubeInformers.InformersFor("").Storage().V1().CSIDrivers().Informer(),
+	).ToController("CSIDriverStarter", eventRecorder)
+}
+
+func (c *CSIDriverStarterControllerHyperShift) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("CSIDriverStarterController.Sync started")
+	defer klog.V(4).Infof("CSIDriverStarterController.Sync finished")
+
+	opSpec, _, _, err := c.guestClient.OperatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	infrastructure, err := c.infraLister.Get(infraConfigName)
+	if err != nil {
+		return err
+	}
+	featureGate, err := c.featureGateLister.Get(featureGateConfigName)
+	if err != nil {
+		return err
+	}
+
+	// Start controller managers for this platform
+	for i := range c.controllers {
+		ctrl := &c.controllers[i]
+
+		csiDriver, err := c.csiDriverLister.Get(ctrl.operatorConfig.CSIDriverName)
+		if errors.IsNotFound(err) {
+			err = nil
+			csiDriver = nil
+		}
+		if err != nil {
+			return err
+		}
+
+		if !ctrl.running {
+			shouldRun, err := shouldRunController(ctrl.operatorConfig, infrastructure, featureGate, csiDriver)
+			if err != nil {
+				return err
+			}
+			if !shouldRun {
+				continue
+			}
+			// add static assets
+			objs, err := ctrl.ctrlRelatedObjects.RelatedObjects()
+			if err != nil {
+				if isNoMatchError(err) {
+					// RESTMapper NoResourceMatch / NoKindMatch errors are cached. Reset the cache to get fresh results on the next sync.
+					c.restMapper.Reset()
+				}
+				return err
+			}
+			relatedObjects = append(relatedObjects, objs...)
+			relatedObjects = append(relatedObjects, configv1.ObjectReference{
+				Group:    operatorapi.GroupName,
+				Resource: "clustercsidrivers",
+				Name:     ctrl.operatorConfig.CSIDriverName,
+			})
+			klog.V(2).Infof("Starting ControllerManager for %s", ctrl.operatorConfig.ConditionPrefix)
+			go ctrl.mgr.Start(ctx)
+			ctrl.running = true
+		}
+	}
+	return nil
+}
+
+func (c *CSIDriverStarterControllerHyperShift) createCSIControllerManager(
```

Comment:
Does it make sense to have two *DriverStarters instead of one? Their code before starting a CSI driver operator looks quite same, the DriverStarter could have just different functions `createStandaloneCSIControllerManager` / `createHyperShiftCSIControllerManager`.

I don't mind running two StaticResourceControllers in the standalone cluster.

Reply:
There are still enough difference between them that, I would like to keep them separate for now until we can think of a cleaner way of reducing duplication in the code. I sort of don't like superfluous conditions being added to a list which is already pretty long.  I sort of prefer some duplication until we have cleaner design. This keeps standalone cluster code almost untouched.